### PR TITLE
MONGOID-5105: Mongoid::Association::EmbedsMany::Proxy#count should allow block form

### PR DESCRIPTION
**Pull request by johnnyshields**
Fixes MONGOID-5105

`Mongoid::Association::EmbedsMany::Proxy#count` should allow block form to make it congruent with #any?, #none?, #all?, etc.

Interestingly this already works for normal relations but not embedded ones.

Example:

```ruby
person.addresses.create(country: "FR")
person.addresses.create(country: "US")
person.addresses.build(country: "FR")

person.addresses.count {|a| a.country == "FR" }  #=> 2
```

To fully be congruent with the `method_missing` behavior, I've also added support for the `args` form of `#count`, although it's not super useful in practice:

```ruby
addr = person.addresses.first

person.addresses.count(addr)  #=> 1
person.addresses.count("foobar")  #=> 0
```

I've also added additional tests for other methods `#any?`, `#none?`, `#all?`

One of the reasons for doing this is that Rubocop/Performance corrects the following:

```
person.addresses.select {|a| a.country == "FR" }.size  # BAD

person.addresses.count {|a| a.country == "FR" }  # GOOD
```

So this will help avoid accidental errors.

### CODE REVIEW

**Comment by johnnyshields** [comment]
Haha I was wondering if you'd notice that!

**Comment by johnnyshields** [comment]
Can this be merged?

**Comment by johnnyshields** [comment]
@comandeo done

**Comment by comandeo** [comment]
@johnnyshields Unfortunately, tests for this PR fail on Ruby 2.3 and 2.4:

```
  1) Mongoid::Association::Embedded::EmbedsMany::Proxy#any? when documents are persisted argument form is supported
      Failure/Error: expect(person.addresses.any?(person.addresses.first)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1439:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'

    2) Mongoid::Association::Embedded::EmbedsMany::Proxy#any? when documents are not persisted argument form is supported
      Failure/Error: expect(person.addresses.any?(person.addresses.first)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1459:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     3) Mongoid::Association::Embedded::EmbedsMany::Proxy#any? when documents are not present argument form is supported
      Failure/Error: expect(person.addresses.any?(1)).to be false
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1474:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     4) Mongoid::Association::Embedded::EmbedsMany::Proxy#all? when documents are persisted argument form is supported
      Failure/Error: expect(person.addresses.all?(person.addresses.first)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1500:in `all?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1500:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     5) Mongoid::Association::Embedded::EmbedsMany::Proxy#all? when documents are not persisted argument form is supported
      Failure/Error: expect(person.addresses.all?(person.addresses.first)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1520:in `all?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1520:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     6) Mongoid::Association::Embedded::EmbedsMany::Proxy#all? when documents are not present argument form is supported
      Failure/Error: expect(person.addresses.all?(nil)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1535:in `all?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1535:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     7) Mongoid::Association::Embedded::EmbedsMany::Proxy#none? when documents are persisted argument form is supported
      Failure/Error: expect(person.addresses.none?(person.addresses.first)).to be false
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1562:in `none?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1562:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'

    8) Mongoid::Association::Embedded::EmbedsMany::Proxy#none? when documents are not persisted argument form is supported
      Failure/Error: expect(person.addresses.none?(person.addresses.first)).to be false
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1582:in `none?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1582:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
     9) Mongoid::Association::Embedded::EmbedsMany::Proxy#none? when documents are not present argument form is supported
      Failure/Error: expect(person.addresses.none?(nil)).to be true
      
      ArgumentError:
        wrong number of arguments (given 1, expected 0)
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1597:in `none?'
      # ./spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1597:in `block (4 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:68:in `block (3 levels) in <top (required)>'
      # ./spec/lite_spec_helper.rb:67:in `block (2 levels) in <top (required)>'
```
Mongoid still supports 2.3 and 2.4 even in the latest version - https://docs.mongodb.com/mongoid/master/reference/compatibility/#ruby-compatibility 
So, these failures should be fixed to get the PR merged.

**Comment by johnnyshields** [comment]
@comandeo thanks. The correct solution here is just to ignore those tests on older Ruby. The mongoid code calls `super` and the older ruby versions don't support some of the syntax like `.any?(1)`

**Comment by johnnyshields** [comment]
@comandeo may I ask why the CI is still failing?

**Comment by comandeo** [comment]
@johnnyshields CI is pretty green now, actually. The only configuration that fails is jruby-app-test; however, this configuration fails everywhere last couple of days. I created a ticked to investigate - https://jira.mongodb.org/browse/MONGOID-5134 - and disabled this configuration temporarily.

So, if you merge the current master into your branch, everything should pass.

**Comment by johnnyshields** [comment]
OK done

**Comment by johnnyshields** [comment]
Now Ruby 3.0 failing?

**Comment by comandeo** [comment]
Actually, not, everything pass. There was a problem with a runner, some connectivity issues. Now all good.

File: lib/mongoid/association/embedded/embeds_many/proxy.rb [hunk 1]
```diff
@@ -107,12 +107,25 @@ def clear
           #
           # Use #size if you want the total number of documents.
           #
+          # If args or block are present, #count will delegate to the target and
```
Comment (p-mongo) [inline lib/mongoid/association/embedded/embeds_many/proxy.rb:110]: #count will delegate to the #count method on +target+ ?

File: spec/mongoid/association/embedded/embeds_many/proxy_spec.rb [hunk 2]
```diff
@@ -1426,12 +1453,186 @@ class TrackingIdValidationHistory
...
+          expect(person.addresses.any?(1)).to be false
+        end
+      end
+    end
+  end
+
+  describe "#all?" do
+
+    let(:person) do
+      Person.create
```
Comment (p-mongo) [inline spec/mongoid/association/embedded/embeds_many/proxy_spec.rb:1494]: These should all use `create!` to verify the creation succeeded.

**Review by p-mongo — COMMENTED** [review]
I believe the behavior of Mongoid on all supported Ruby versions should be the same, if practical. I believe this generally how Rails works.

Therefore I believe either this PR would need to implement the missing block form behavior on Ruby 2.3 and 2.4, or Mongoid should increase the minimum required version to 2.5 at which point the 2.5+ checks in this PR would be unneeded.

In https://github.com/mongodb/mongo-ruby-driver/pull/2298 we increased the minimum Ruby version to 2.4 for the driver, but per the docs in https://docs.mongodb.com/mongoid/master/reference/compatibility/#ruby-compatibility 2.3 and 2.4 support has already been deprecated for two Mongoid releases and perhaps the minimum can be increased to 2.5 to save work in this PR.

**Comment by johnnyshields** [comment]
> I believe the behavior of Mongoid on all supported Ruby versions should be the same, if practical.
The behavior on all versions is the same. We delegate `count` etc. to the underlying object. The underlying object behavior may different on Ruby versions. It is inadvisable to add a monkey patch here on Ruby kernel methods; I'm not going to change this aspect of my PR.

**Comment by p-mongo** [comment]
No patching of standard library methods is needed, the needed code can be added to Mongoid methods based on active Ruby version.

**Comment by johnnyshields** [comment]
Not worth the effort; I'm not going to add Ruby 2.3 and 2.4 specific hacks in the Mongoid codebase.

File: spec/mongoid/association/embedded/embeds_many/proxy_spec.rb (hunk 2 above)
Comment (johnnyshields) [reply spec/mongoid/association/embedded/embeds_many/proxy_spec.rb]: Done. I've changed all usages of .create in the spec file.

File: lib/mongoid/association/embedded/embeds_many/proxy.rb (hunk 1 above)
Comment (johnnyshields) [reply lib/mongoid/association/embedded/embeds_many/proxy.rb]: Done.

**Comment by johnnyshields** [comment]
Shall we merge this? If Mongoid team would like to add the hacks it can be done as a separate PR; be aware they apply to many methods (`#any?`, `#all?`, `#none?`, etc.) which are not being touched in this PR (this PR only affects `#count`)

**Comment by dblock** [comment]
I am confused how the build is green but more work is required to get this PR merged?

**Comment by johnnyshields** [comment]
@comandeo looks like this needs your review?

**Comment by p-mongo** [comment]
#5060

**Comment by johnnyshields** [comment]
Closing in favor of #5060
